### PR TITLE
Add more fields to user

### PR DESCRIPTION
**Pull request by romiparada**
https://trello.com/c/Hsh74P2l

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -3,4 +3,19 @@
...
+  before_action :require_user_fields, unless: :devise_controller?
+
+  private
+
+  def require_user_fields
+    user = current_user
+    missing_fileds = %i[name birthdate username].select do |field|
+      user.send(field).blank?
+    end
+    return if missing_fileds.blank?
```
Comment (micaelalorenzo) [inline app/controllers/application_controller.rb:16]: I think this approach is a little complex but I like you show the attributes that are missing. May be Apa can give you a more accurate review. I think I would just do a single line like `return if user.name.present? && ... `. You can also delegate this to a method in the user model.

File: app/controllers/application_controller.rb [hunk 2]
```diff
@@ -3,4 +3,16 @@
 class ApplicationController < ActionController::API
   include ActionController::MimeResponds
   respond_to :json
+  before_action :authenticate_user!
+  before_action :require_user_fields, unless: :devise_controller?
```
Comment (juan-apa) [inline app/controllers/application_controller.rb:7]: let's rename the method to something that sounds more business-like. e.g. `user_profile_filled`

File: app/models/user.rb [hunk 3]
```diff
@@ -6,4 +6,25 @@ class User < ApplicationRecord
...
+  validates :website, format: { with: %r{https?://(www\.)?(.*)\.com} }, allow_blank: true
+  validates :username, uniqueness: { case_sensitive: false }, length: { in: 2..20 }, allow_blank: true
+
+  validate :over_eighteen
+
+  def missing_require_fields
+    %i[name birthdate username].select do |field|
+      send(field).blank?
+    end
+  end
```
Comment (juan-apa) [inline app/models/user.rb:21]: there's a typo in the method name: `missing_required_fields`

File: app/models/user.rb [hunk 4]
```diff
@@ -6,4 +6,25 @@ class User < ApplicationRecord
   devise :database_authenticatable, :registerable, :confirmable,
          :recoverable, :rememberable, :validatable,
          :jwt_authenticatable, jwt_revocation_strategy: self
+
+  validates :name, length: { minimum: 2 }, allow_blank: true
+  validates :bio, length: { maximum: 160 }, allow_blank: true
+  validates :website, format: { with: %r{https?://(www\.)?(.*)\.com} }, allow_blank: true
```
Comment (juan-apa) [inline app/models/user.rb:12]: you can use `URI::DEFAULT_PARSER.make_regexp(%w[http https])` instead of the regex

File: app/models/user.rb [hunk 5]
```diff
@@ -6,4 +6,25 @@ class User < ApplicationRecord
...
+  validates :username, uniqueness: { case_sensitive: false }, length: { in: 2..20 }, allow_blank: true
+
+  validate :over_eighteen
```
Comment (juan-apa) [inline app/models/user.rb:15]: let's rename the method to `over_18_years_old`, which is self explanatory

File: spec/factories/users.rb [hunk 6]
```diff
@@ -6,6 +6,15 @@
...
 
+    name { Faker::Internet.username(specifier: 2) }
+    bio { Faker::Lorem.paragraph_by_chars(number: 160) }
+    website do
+      ["http://#{Faker::Internet.domain_word}.com", "http://www.#{Faker::Internet.domain_word}.com",
+       "https://#{Faker::Internet.domain_word}.com", "https://#{Faker::Internet.domain_word}.com"].sample
+    end
+    birthdate { Faker::Date.birthday(min_age: 18) }
+    username { Faker::Internet.username(specifier: 2..20) }
+
```
Comment (juan-apa) [inline spec/factories/users.rb:17]: can you make a trait for this?
```ruby
trait :without_profile_data do
  name {}
  ....
end
```

File: spec/models/user_spec.rb [hunk 7]
```diff
@@ -6,8 +6,38 @@
...
+
+    context 'validates website format' do
+      it { is_expected.to allow_value('http://www.valid.com').for(:website) }
+      it { is_expected.to allow_value('https://www.valid.com').for(:website) }
+      it { is_expected.to allow_value('http://valid.com').for(:website) }
+      it { is_expected.to allow_value('https://valid.com').for(:website) }
+      it { is_expected.to_not allow_value(Faker::Internet.domain_name).for(:website) }
+    end
+
+    context 'validates birthday' do
```
Comment (juan-apa) [inline spec/models/user_spec.rb:39]: use a `describe '#over_18_years_old'`

Comment (romiparada) [reply spec/models/user_spec.rb:39]: over_18_years_old is a private method and adds to user.errors when the user is under 18. 
Would it be correct to change it to?

```
  describe '#over_18_years_old' do
    context 'when user is under 18' do
      before { subject.birthdate = Faker::Date.birthday(max_age: 17) }

      it 'adds errors to user' do
        subject.send(:over_18_years_old)
        expect(subject.errors[:birthdate]).to match('User must be at least 18 years old')
      end
    end
  end
```

**Review by juan-apa — APPROVED** [review]

**Pull request merged**